staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go

```diff
@@ -85,9 +85,9 @@ func Register() {
 // uppercase to be backwards compatible with existing monitoring tooling.
 func Monitor(verb, resource, subresource, scope, client, contentType string, httpCode, respSize int, reqStart time.Time) {
 	elapsed := float64((time.Since(reqStart)) / time.Microsecond)
-	requestCounter.WithLabelValues(verb, resource, subresource, client, contentType, codeToString(httpCode)).Inc()
-	requestLatencies.WithLabelValues(verb, resource, subresource).Observe(elapsed)
-	requestLatenciesSummary.WithLabelValues(verb, resource, subresource).Observe(elapsed)
+	requestCounter.WithLabelValues(verb, resource, subresource, scope, client, contentType, codeToString(httpCode)).Inc()
```

Comment:
This metric count is already unbounded due to client label, adding scope will now triple them. Discussion with @wojtek-t about this earlier - https://github.com/kubernetes/kubernetes/pull/49117#issuecomment-316702136
Do we still want to do this?

Reply:
Only if the client actually invokes all three scopes.  Which is rarely the case.

I checked on our largest cluster (20k+ unique users and namespaces, not as many nodes).  I think this fairly high cardinality generating, but much smaller than the other metrics.  Today:

```
# sort_desc(count by (__name__)({__name__=~"apiserver_.*"}))
apiserver_request_latencies_bucket{} | 6184
apiserver_request_count{} | 4501
apiserver_request_latencies_summary{} | 2319
apiserver_request_latencies_summary_count{} | 773
apiserver_request_latencies_count{} | 773
apiserver_request_latencies_summary_sum{} | 773
apiserver_request_latencies_sum{} | 773
```

The highest api server metric cardinality is about 60 items down the list:

```
container_tasks_state{} | 229635
container_cpu_usage_seconds_total{} | 196684
container_memory_failures_total{} | 183708
container_fs_writes_total{} | 159961
container_fs_reads_total{} | 159961
container_fs_writes_bytes_total{} | 142592
container_fs_reads_bytes_total{} | 142592
container_memory_cache{} | 45927
container_memory_rss{} | 45927
container_memory_working_set_bytes{} | 45927
container_start_time_seconds{} | 45927
container_memory_failcnt{} | 45927
container_cpu_system_seconds_total{} | 45927
container_memory_usage_bytes{} | 45927
container_last_seen{} | 45927
container_cpu_user_seconds_total{} | 45927
container_memory_swap{} | 45927
container_spec_cpu_shares{} | 45873
container_spec_cpu_period{} | 45873
container_spec_memory_swap_limit_bytes{} | 45732
container_spec_memory_limit_bytes{} | 45732
container_fs_sector_reads_total{} | 18164
container_fs_write_seconds_total{} | 18164
container_fs_inodes_free{} | 18164
container_fs_io_current{} | 18164
container_fs_io_time_weighted_seconds_total{} | 18164
container_fs_reads_merged_total{} | 18164
container_fs_writes_merged_total{} | 18164
container_fs_inodes_total{} | 18164
container_fs_read_seconds_total{} | 18164
container_fs_sector_writes_total{} | 18164
container_fs_io_time_seconds_total{} | 18164
container_fs_usage_bytes{} | 18164
container_fs_limit_bytes{} | 18164
kubelet_runtime_operations_latency_microseconds{} | 13146
container_network_transmit_bytes_total{} | 11457
container_network_transmit_errors_total{} | 11457
container_network_transmit_packets_total{} | 11457
container_network_transmit_packets_dropped_total{} | 11457
container_network_receive_bytes_total{} | 11457
container_network_receive_packets_dropped_total{} | 11457
container_network_receive_errors_total{} | 11457
container_network_receive_packets_total{} | 11457
kubelet_docker_operations_latency_microseconds{} | 8649
container_cpu_cfs_throttled_periods_total{} | 8245
container_spec_cpu_quota{} | 8245
container_cpu_cfs_throttled_seconds_total{} | 8245
container_cpu_cfs_periods_total{} | 8245
node_dns_response_size_bytes_bucket{} | 6622
```

Total samples scraped

```
# sum by (job) (scrape_samples_scraped)
{job="kubernetes-nodes"} | 2863747
{job="kubernetes-apiservers"} | 17074
```

So yes, it will increase cardinality.  But I think this cardinality is more useful than client cardinality, and in practice we should we be better off limiting that unbounded cardinality than not gathering this metric.

Reply:
That sounds reasonable.. thanks for the info.